Enable D-separation on counterfactual graphs

It turns out we need to check for D-separation in the counterfactual graph in order to fix the flaw in the ID* algorithm. However, if we try to do so, we currently get this error when running 
[test_id_star.py](https://github.com/y0-causal-inference/y0/blob/17d79830975ff69236edf9c268e5a2df991b3ede/tests/test_algorithm/test_id_star.py#LL322C1-L344C133)

```python

test_id_star.py:344:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../src/y0/algorithm/identify/id_star.py:360: in simplify_counterfactual
    if are_d_separated(
../../src/y0/algorithm/conditional_independencies.py:165: in are_d_separated
    admg = graph.to_admg()
../../src/y0/graph.py:120: in to_admg
    self.raise_on_counterfactual()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ self = NxMixedGraph(directed=<networkx.classes.digraph.DiGraph object at 0x7f7839123df0>, undirected=<networkx.classes.graph.Graph object at 0x7f7839123e80>)
    def raise_on_counterfactual(self) -> None:
    """Raise an error if this is a counterfactual graph.
    :raises ValueError: if this graph is a counterfactual graph
    """
    if self.is_counterfactual():
>           raise ValueError("This operation is not available for counterfactual graphs")
E           ValueError: This operation is not available for counterfactual graphs
../../src/y0/graph.py:90: ValueError'


Edited<https://teams.microsoft.com/l/message/19:3efe1b0be7b14e718c36f5cfe71dd95c@thread.tacv2/1683000130755?tenantId=d6faa5f9-0ae2-4033-8c01-30048a38deeb&amp;groupId=940a4654-27a9-4a94-a146-e052b85ba421&amp;parentMessageId=1683000130755&amp;teamName=KESTREL&amp;channelName=Task 3&amp;createdTime=1683000130755&amp;allowXTenantAccess=false>
```

--- a/src/y0/algorithm/conditional_independencies.py
+++ b/src/y0/algorithm/conditional_independencies.py
@@ -107,7 +107,6 @@ def disorient(graph: NxMixedGraph) -> nx.Graph:
     """Convert an :mod:`ananke` mixed directed/undirected into a undirected (networkx) graph.
 
     :param graph: An ananke graph
-    :type graph: ananke.graphs.SG
     :returns: A disoriented graph
     """
     rv = nx.Graph()
@@ -123,7 +122,6 @@ def get_moral_links(graph: NxMixedGraph) -> List[Tuple[Variable, Variable]]:
     May generate links that already exist as we assume we are not working on a multi-graph.
 
     :param graph: Graph to process
-    :type graph: ananke.graphs.SG
     :return: An collection of edges to add.
     """
     parents = [graph.directed.predecessors(node) for node in graph.nodes()]

--- a/src/y0/algorithm/conditional_independencies.py
+++ b/src/y0/algorithm/conditional_independencies.py
@@ -103,7 +103,7 @@ def _len_lex(judgement: DSeparationJudgement) -> Tuple[int, str]:
     return len(judgement.conditions), ",".join(c.name for c in judgement.conditions)
 
 
-def disorient(graph) -> nx.Graph:
+def disorient(graph: NxMixedGraph) -> nx.Graph:
     """Convert an :mod:`ananke` mixed directed/undirected into a undirected (networkx) graph.
 
     :param graph: An ananke graph
@@ -111,12 +111,13 @@ def disorient(graph) -> nx.Graph:
     :returns: A disoriented graph
     """
     rv = nx.Graph()
-    rv.add_nodes_from(graph.vertices)
-    rv.add_edges_from(chain(graph.di_edges, graph.ud_edges, graph.bi_edges))
+    rv.add_nodes_from(graph.nodes())
+    rv.add_edges_from(graph.directed.edges())
+    rv.add_edges_from(graph.undirected.edges())
     return rv
 
 
-def get_moral_links(graph) -> List[Tuple[Variable, Variable]]:
+def get_moral_links(graph: NxMixedGraph) -> List[Tuple[Variable, Variable]]:
     """Generate links to ensure all co-parents in a graph are linked.
 
     May generate links that already exist as we assume we are not working on a multi-graph.
@@ -125,7 +126,7 @@ def get_moral_links(graph) -> List[Tuple[Variable, Variable]]:
     :type graph: ananke.graphs.SG
     :return: An collection of edges to add.
     """
-    parents = [graph.parents([v]) for v in graph.vertices]
+    parents = [graph.directed.predecessors(node) for node in graph.nodes()]
     moral_links = [*chain(*[combinations(nodes, 2) for nodes in parents if len(parents) > 1])]
     return moral_links
 
@@ -159,27 +160,25 @@ def are_d_separated(
     if not all(isinstance(c, Variable) for c in conditions):
         raise TypeError(f"some conditions are not variables: {conditions}")
 
-    condition_names = {c.name for c in conditions}
-    named = {a.name, b.name}.union(condition_names)
-
-    admg = graph.to_admg()
+    condition_names = {c for c in conditions}
+    named = {a, b}.union(condition_names)
 
     # Filter to ancestors
-    keep = admg.ancestors(named)
-    admg = copy.deepcopy(admg.subgraph(keep))
+    keep = graph.ancestors_inclusive(named)
+    sg = copy.deepcopy(graph.subgraph(keep))
 
     # Moralize (link parents of mentioned nodes)
-    for u, v in get_moral_links(admg):  # type: ignore
-        admg.add_udedge(u, v)
+    for u, v in get_moral_links(sg):
+        sg.add_undirected_edge(u, v)
 
     # disorient & remove conditions
-    evidence_graph = disorient(admg)
+    evidence_graph = disorient(sg)
 
     keep = set(evidence_graph.nodes) - set(condition_names)
     evidence_graph = evidence_graph.subgraph(keep)
 
     # check for path....
-    separated = not nx.has_path(evidence_graph, a.name, b.name)  # If no path, then d-separated!
+    separated = not nx.has_path(evidence_graph, a, b)  # If no path, then d-separated!
 
     return DSeparationJudgement.create(left=a, right=b, conditions=conditions, separated=separated)
 

--- a/src/y0/graph.py
+++ b/src/y0/graph.py
@@ -120,6 +120,8 @@ class NxMixedGraph:
         self.raise_on_counterfactual()
         from ananke.graphs import ADMG
 
+        # update the way stringification happens so this
+        # can support arbitrary variables, like counterfactuals
         return ADMG(
             vertices=[n.name for n in self.nodes()],
             di_edges=[(u.name, v.name) for u, v in self.directed.edges()],

--- a/tests/test_algorithm/test_conditional_independencies.py
+++ b/tests/test_algorithm/test_conditional_independencies.py
@@ -44,9 +44,14 @@ class TestDSeparation(unittest.TestCase):
         for example in testable:
             with self.subTest(name=example.name):
                 for ci in example.conditional_independencies:
+                    self.assertIn(ci.left, example.graph)
+                    self.assertIn(ci.right, example.graph)
+                    judgement = are_d_separated(
+                        example.graph, ci.left, ci.right, conditions=ci.conditions
+                    )
                     self.assertTrue(
-                        are_d_separated(example.graph, ci.left, ci.right, conditions=ci.conditions),
-                        msg="Expected d-separation not found",
+                        judgement,
+                        msg=f"Expected d-separation not found in {example.name}",
                     )
                     if ci.conditions:
                         self.assertFalse(
@@ -59,49 +64,49 @@ class TestDSeparation(unittest.TestCase):
 
         This test covers several cases around moral links to ensure that they are added when needed.
         """
-        from ananke.graphs import ADMG
-
-        graph = ADMG(
-            vertices=("a", "b", "c"),
-            di_edges=[("a", "b"), ("b", "c")],
+        graph = NxMixedGraph.from_str_edges(
+            nodes=("a", "b", "c"),
+            directed=[("a", "b"), ("b", "c")],
         )
-        links = set(tuple(sorted(e)) for e in get_moral_links(graph))
-        self.assertEqual(set(), links, msg="Unexpected moral links added.")
+        links = list(get_moral_links(graph))
+        self.assertEqual([], links, msg="Unexpected moral links added.")
 
-        graph = ADMG(
-            vertices=("a", "b", "c"),
-            di_edges=[("a", "c"), ("b", "c")],
+        graph = NxMixedGraph.from_str_edges(
+            nodes=("a", "b", "c"),
+            directed=[("a", "c"), ("b", "c")],
         )
         links = set(tuple(sorted(e)) for e in get_moral_links(graph))
         self.assertEqual(
-            {("a", "b")}, links, msg="Moral links not as expected in single-link case."
+            {(Variable("a"), Variable("b"))},
+            links,
+            msg="Moral links not as expected in single-link case.",
         )
 
-        graph = ADMG(
-            vertices=("a", "b", "aa", "bb", "c"),
-            di_edges=[("a", "c"), ("b", "c"), ("aa", "c"), ("bb", "c")],
+        graph = NxMixedGraph.from_str_edges(
+            nodes=("a", "b", "aa", "bb", "c"),
+            directed=[("a", "c"), ("b", "c"), ("aa", "c"), ("bb", "c")],
         )
         links = set(tuple(sorted(e)) for e in get_moral_links(graph))
         self.assertEqual(
             {
-                ("a", "b"),
-                ("a", "aa"),
-                ("a", "bb"),
-                ("aa", "b"),
-                ("aa", "bb"),
-                ("b", "bb"),
+                (Variable("a"), Variable("b")),
+                (Variable("a"), Variable("aa")),
+                (Variable("a"), Variable("bb")),
+                (Variable("aa"), Variable("b")),
+                (Variable("aa"), Variable("bb")),
+                (Variable("b"), Variable("bb")),
             },
             links,
             msg="Moral links not as expected in multi-link case.",
         )
 
-        graph = ADMG(
-            vertices=("a", "b", "c", "d", "e"),
-            di_edges=[("a", "c"), ("b", "c"), ("c", "e"), ("d", "e")],
+        graph = NxMixedGraph.from_str_edges(
+            nodes=("a", "b", "c", "d", "e"),
+            directed=[("a", "c"), ("b", "c"), ("c", "e"), ("d", "e")],
         )
         links = set(tuple(sorted(e)) for e in get_moral_links(graph))
         self.assertEqual(
-            {("a", "b"), ("c", "d")},
+            {(Variable("a"), Variable("b")), (Variable("c"), Variable("d"))},
             links,
             msg="Moral links not as expected in multi-site case.",
         )

